### PR TITLE
[SPARK-14480][SQL] Remove meaningless StringIteratorReader for CSV data source.

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR removes meaningless `StringIteratorReader` for CSV data source.

In `CSVParser.scala`, there is an `Reader` wrapping `Iterator` but there are two problems by this.

Firstly, it was actually not faster than processing line by line with Iterator due to additional logics to wrap `Iterator` to `Reader`.
Secondly, this brought a bit of complexity because it needs additional logics to allow every line to be read bytes by bytes. So, it was pretty difficult to figure out issues about parsing, (eg. SPARK-14103).

A benchmark was performed manually and the results were below:
- Original codes with Reader wrapping Iterator

| End-to-end (ns) | Parse Time (ns) |
| --- | --- |
| 14116265034 | 2008277960 |
- New codes with Iterator

| End-to-end (ns) | Parse Time (ns) |
| --- | --- |
| 13451699644 | 1549050564 |

For the details for the environment, dataset and methods, please refer the JIRA ticket.
## How was this patch tested?

Existing tests should cover this.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVParser.scala [hunk 1]
```diff
@@ -28,11 +28,10 @@ import org.apache.spark.internal.Logging
  * Read and parse CSV-like input
  *
  * @param params Parameters object
- * @param headers headers for the columns
  */
-private[sql] abstract class CsvReader(params: CSVOptions, headers: Seq[String]) {
+private[sql] class CsvReader(params: CSVOptions) {
 
-  protected lazy val parser: CsvParser = {
+  private lazy val parser: CsvParser = {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVParser.scala]: why is this lazy?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVParser.scala]: Oh, it seems it doesn't have to be. I will remove this and take a close look again.

**Comment by HyukjinKwon** [comment]
@rxin Could you take another look please?

**Comment by rxin** [comment]
This looks good. Let me merge it.

**Comment by davies** [comment]
@HyukjinKwon @rxin This patch have a regression: A column that have escaped newline can't be correctly parsed anymore. Should we revert this patch or figure a way to fix that?

**Comment by HyukjinKwon** [comment]
@davies, removed `StringIteratorReader` concatenates the lines in each iterator into reader in each partition IIRC.

New line in the column was not supported correctly up to my understanding because rows can spawn across multiple blocks. This is a similar problem that we have not supported multiple JSON lines before up to my knowledge. 

Currently,  we have some open PRs for dealing with multiple lines support by using something like `wholeTextFile` or dealing with each file as a multiple line json, which I think we could solve in that way if any of it is merged.

I guess we introduced several regression or behaviour changes when we porting. Would this be acceptable rather than supporting multiple lines one in this way?

**Comment by HyukjinKwon** [comment]
FWIW, I remember I had a hard time to figure out https://issues.apache.org/jira/browse/SPARK-14103 where the issue itself was about quote but it ended up reading whole partition as a value.

**Pull request closed without merging**